### ISSUE TITLE
GroupBy() LINQ method reporting not supported.

### ISSUE THREAD

**Issue by deokershesh** [dotnet/efcore#21464]
There are similar issues opened #17068 and #18102. However I didn't find any solution mentioned! May be I missed it! If so, kindly provide the solution/suggestion. Problem statement is below. I am using EF Core v3.1.

var data = schoolContext.Students.GroupBy(a => a.StudentId);

The above line reports error "Client side GroupBy is not supported." Not sure, why EF Core is not sending the underlying SQL query to database server! instead chooses to process on its own and reports error!

However below direct SQL query works, which confirms underlying database supports the GROUP BY clause.

var _students = schoolContext.Students.FromSqlRaw("SELECT StudentId, GradeID, count(*), Name FROM students GROUP BY StudentId, GradeID, Name;");

Any pointers would be helpful.

**Comment by roji** [comment]
@deokershesh you probably want to take a look at [the docs on GroupBy](https://docs.microsoft.com/en-us/ef/core/querying/complex-query-operators#groupby). A GroupBy without a projection containing an aggregate function (Count) in the example above isn't translatable to SQL. Your raw SQL above contains `COUNT(*)`, but your LINQ query doesn't.

**Comment by deokershesh** [comment]
Will take a look at the doc, however server GROUP BY query works without count(*) also.

**Comment by roji** [comment]
@deokershesh can you post a sample SQL query of a GROUP BY without an aggregate (e.g. count(*))?

**Comment by deokershesh** [comment]
var _students2 = schoolContext.Students.FromSqlRaw("Select StudentId, GradeID, Name from students group by StudentId, GradeID, Name;");

I could see the above query hits the Informix server correctly and runs successfully and returns result set also.

Sess       SQL            Current            Iso Lock       SQL  ISAM F.E.
Id         Stmt type      Database           Lvl Mode       ERR  ERR  Vers  Explain
350        -              sheshefcoredb1     CR  Not Wait   0    0    4.50  Off

Last parsed SQL statement :
  Select StudentId, GradeID, Name from students group by StudentId, GradeID, Name;

**Comment by roji** [comment]
This works because you're grouping by all the columns - I'm really not sure what the utility of that is. Try to remove one of the columns from the GROUP BY clause, and it should fail without an aggregate function.

**Comment by deokershesh** [comment]
I was trying to answer the question by removing count(*), in this case probably it doesn't any value!
Now, I tried, below query by following the document, while it doesn't throw any exception, it doesn't return any success either, the Message member ("query->Results View->Message") says "Specified cast is not valid.". Not sure, what's wrong and how to diagnose the same, as it doesn't generate any SQL query which can be run on the database server!

                var query = from p in schoolContext.Set<Student>()
                            group p by p.StudentId into g
                            select new
                            {
                                g.Key,
                                Count = g.Count()
                            };

**Comment by roji** [comment]
I assume you are trying this on the Informix provider you are building... A good start would be to look at the stack trace from the InvalidCastException which you are seeing. The above is standard and should produce SQL that will run on any relational database.

**Comment by deokershesh** [comment]
Tried adding all possible breakpoints, it doesn't step into the code where the stack shows! Again, as such there is no exception, the below code executes and shows following stack trace in the "query" object members. Adding breakpoint at any of the files/functions below doesn't stop debugger!

var query = from p in schoolContext.Set<Student>()
                        group p by p.StudentId into g
                        select new
                        {
                            g.Key,
                            Count = g.Count()
                        };

   at Informix.Net.Core.IfxDataReader.GetInt32(Int32 i) in D:\corefx-release-3.1\src\Informix.Net.Core\src\Informix\Net\Core\OdbcDataReader.cs:line 1515
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext() in D:\EntityFrameworkCore\efcore-master\src\EFCore.Relational\Query\Internal\QueryingEnumerable.cs:line 202
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at System.Linq.SystemCore_EnumerableDebugView`1.get_Items()

**Comment by roji** [comment]
@deokershesh are you actually evaluating the query? According to your code above, you're missing a ToList (or similar mechanism) to actually make the query evaluate...

**Comment by deokershesh** [comment]
Thanks @roji, After adding ToList() it correctly converts to SQL query which is executed by Informix server correctly.
However on the EF Core provider code, somehow COUNT(*) is being treated as INTEGER instead of DECIMAL (Informix maps COUNT(*) value/type to DECIMAL). This causes the InvalidCastException. 

Basically the below code in (public bool MoveNext() method) ....\EFCore.Relational\Query\QueryEnumerable.cs file calls the GetInt32() method of IfxDataReader class of Informix .NET Core Provider. When it moves to COUNT(*) field since its DECIMAL type for Informix, EF Core continues to call GetInt32() method instead of GetDecimal() method. I believe somewhere mapping is missing in the EF Core code which I need to change for such use cases for Informix. If you have an idea about such mapping to be done in which part of EF Code code and could let me know would be great help.

Current = _shaper(
                                    _relationalQueryContext, _dataReader.DbDataReader,
                                    _resultCoordinator.ResultContext, _indexMap, _resultCoordinator);

**Comment by deokershesh** [comment]
Fixed the issue, there is TranslateCount() method in ....
\EFCore.Relational\Query\RelationalSqlTranslatingExpressionVisitor.cs file. Which translates COUNT( * )  to native database type. This was typeof(int) [I believe SQL Server treats this as INTEGER type], which I had to change to typeof(decimal) and it all worked pretty well.

Thanks @roji for pointers....

Similarly this file ....\EFCore.Relational\Query\RelationalSqlTranslatingExpressionVisitor.cs has other aggregate function handlings like MIN(), MAX(), SUM() etc. Which may need appropriate handling for Informix database.
